File: leetcode/golang/medium/Problem33.go
package main

import (
	"fmt"
)

// Requirement: O(log n) performance
// Hint:
func search(nums []int, target int) int {
	low, high := 0, len(nums)-1
	for low <= high {
		mid := low + (high-low)/2
		if nums[mid] == target {
			return mid
		}

		if nums[low] <= nums[mid] {
			if nums[low] <= target && target < nums[mid] {
				high = mid - 1
			} else {
				low = mid + 1
			}
		} else {
			// Large numbers are on left side
			if nums[mid] < target && target <= nums[high] {
				low = mid + 1
			} else {
				high = mid - 1
			}
		}
	}

	return -1
}

// Leetcode 33
// Link: https://leetcode.com/problems/search-in-rotated-sorted-array/description/
func main() {
	fmt.Println(search([]int{4, 5, 6, 7, 0, 1, 2}, 0))
	fmt.Println(search([]int{4, 5, 6, 7, 0, 1, 2}, 3))
	fmt.Println(search([]int{1}, 0))
}

File: leetcode/python/medium/Problem33.py
# Questions to ask:
# 1. What is the time complexity? O(log n)
# 2. What is the space complexity? O(1)
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums)-1
        while l <= r:
            mid = (r+l)//2
            if nums[mid] == target:
                return mid

            if nums[l] <= nums[mid]:
                if nums[l] <= target < nums[mid]:
                    r = mid - 1
                else:
                    l = mid + 1
            else:
                if nums[mid] < target <= nums[r]:
                    l = mid + 1
                else:
                    r = mid - 1
        return -1


# Problem 33
# Link: https://leetcode.com/problems/search-in-rotated-sorted-array/description/
if __name__ == '__main__':
    s = Solution()
    cases = [
        ([4,5,6,7,0,1,2], 0, 4),
        ([4,5,6,7,0,1,2], 3, -1)
    ]
    for nums, target, expected in cases:
        assert s.search(nums, target) == expected
